text/3503-frontmatter.md

```diff
@@ -0,0 +1,526 @@
+  - Visual clutter (where clutter is overwhelming already in Rust)
+- Might be a bit complicated to do edits (translating between location within
+  `toml_edit` spans to the location within `syn` spans especially with different comment styles)
+- Either we expose `syn`s lesser parse errors or we skip errors, deferring to rustc's, but then have the wrong behavior on commands that don't invoke rustc, like `cargo metadata`
+- Requires pulling in a full markdown parser to extract the manifest
+  - Incorrectly formatted markdown would lead to a missing manifest and confusing error messages at best or silent incorrect behavior at worse
+
+### Alternative 2: Macro
+
+```rust
+#!/usr/bin/env cargo
+
+cargo! {
+[package]
+edition = "2018"
+}
+
+fn main() {
+}
+```
+Benefits
+- Parsers are available to make this work (e.g. `syn`)
+
+Downsides
+- When discussing with a Rust crash course teacher, it was felt their students would have a hard time learning to write these manifests from scratch
+  - Unpredictable location (both the doc comment and the cargo code block within it)
+- The `cargo` macro would need to come from somewhere (`std`?) which means it is taking on `cargo`-specific knowledge
+  - An unexplored direction we could go with this is a `meta!` macro (e.g. we'd need to have a format marker in it)
+- A lot of tools/IDEs have problems in dealing with macros
+- Free-form rust code makes it harder for cargo to make edits to the manifest
+- Either we expose `syn`s lesser parse errors or we skip errors, deferring to rustc's, but then have the wrong behavior on commands that don't invoke rustc, like `cargo metadata`
+
+### Alternative 3: Attribute
+
+```rust
+#!/usr/bin/env cargo
+
+#![cargo(manifest = r#"
+[package]
+edition = "2018"
+"#)]
+
+fn main() {
+}
+```
+- `cargo` could register this attribute or `rustc` could get a generic `metadata` attribute
+- As an alternative, `manifest` could a less stringly-typed format but that
+  makes it harder for cargo to parse and edit, makes it harder for users to
+  migrate between single and multi-file packages, and makes it harder to transfer
+  knowledge and experience
+
+Benefits
+- Parsers are available to make this work (e.g. `syn`)
+
+Downsides
+- When discussing with a Rust crash course teacher, it was felt their students would have a hard time learning to write these manifests from scratch
+  - Unpredictable location (both the doc comment and the cargo code block within it)
+- From talking to a teacher, users are more forgiving of not understanding the details for structure data in an unstructured format (doc comments / comments) but something that looks meaningful, they will want to understand it all requiring dealing with all of the concepts
+ - The attribute approach requires explaining multiple "advanced" topics: One teacher doesn't get to teaching any attributes until the second level in his crash course series and two teachers have found it difficult to teach people raw strings
```

Comment:
I don't fully buy into this objection: not everything has to be explained as a pre-requisite to it being used in a course.
Let's take `println!' as an example: [the Rust book](https://doc.rust-lang.org/book/ch01-02-hello-world.html#anatomy-of-a-rust-program) introduces `println!` in the first chapter but it doesn't provide any macro discussion beyond "`println!` is a macro, we'll talk about that later", where "later" is in chapter 19 (!).

I don't see what makes a `cargo` attribute any different here from `println!` or a `#[derive(Debug)]`.

Reply:
> Let's take println! as an example: [the Rust book](https://doc.rust-lang.org/book/ch01-02-hello-world.html#anatomy-of-a-rust-program) introduces println! in the first chapter but it doesn't provide any macro discussion beyond "println! is a macro, we'll talk about that later", where "later" is in chapter 19 (!).

While I'm speaking for myself and not that person, I feel there is a big difference between `println!` and 
```rust
#![cargo(manifest = r#"
[package]
edition = "2018"
"#)]
```

With `println!` the name (mostly) makes sense and it was a weird `!` after it that can be glossed over.  Thats less the case with attributes.

> Elaborating further: aren't we over-indexing on language newcomers here?

I suspect we under-index on language newcomers.

That aside, one of the big use cases for this specific feature is helping users of all levels figure out how to write the code they need, including looking at written material (books, blogs, messages from coworkers, etc).

Reply:
I'm not saying we should discount newcomers entirely, but 

> including looking at written material (books, blogs, messages from coworkers, etc).

mostly involves copy-pasting snippets. After looking at one or two examples they should become accustomed to the syntax and able to use it, regardless of them understanding the ins and outs of attributes.

Reply:
Yes, just teach this as a magic syntax at first. Later on people will understand what it means.